server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.util.timer;
+
+public abstract class TimerTask implements Runnable {
+    private volatile TimerTaskEntry timerTaskEntry;
+    // timestamp in millisecond
+    public final long delayMs;
+
+    public TimerTask(long delayMs) {
+        this.delayMs = delayMs;
+    }
+
+    public void cancel() {
+        synchronized (this) {
+            if (timerTaskEntry != null) timerTaskEntry.remove();
+            timerTaskEntry = null;
+        }
+    }
+
+    void setTimerTaskEntry(TimerTaskEntry entry) {
+        synchronized (this) {
+            // if this timerTask is already held by an existing timer task entry,
+            // we will remove such an entry first.
+            if (timerTaskEntry != null && !timerTaskEntry.equals(entry)) {
```

Comment:
@divijvaidya I was wondering if using `equals` is correct here. It was like this in Scala but as we compare references in all other places, it is a bit weird. Practically, it does not make any difference because `equals` is not overridden anyway.

Reply:
yes, it should be a reference comparison. There can be two entries in the list which are same i.e. have same expirationMs and other members but task could be held by only one of them.

Good observation here.

Reply:
Fixed in last commit.